config: add index_cache_fraction

michoecho opened the pull request:
Index caching was disabled by default because it caused performance regressions
for some small-partition workloads. See https://github.com/scylladb/scylladb/issues/11202.

However, it also means that there are workloads which could benefit from the
index cache, but (by default) don't.

As a compromise, we can set a default limit on the memory usage of index cache,
which should be small enough to avoid catastrophic regressions in
small-partition workloads, but big enough to accommodate workloads where
index cache is obviously beneficial.

This series adds such a configurable limit, sets it to to 0.2 of total cache memory by default,
and re-enables index caching by default.

Fixes #15118

michoecho commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2946/

Huh, TIL that copy initialization doesn't consider explicit constructors.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2948/

bhalevy commented:
@tgrabiec please review

avikivity commented on db/config.cc:
```diff
@@ -1001,7 +1001,9 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , cache_index_pages(this, "cache_index_pages", liveness::LiveUpdate, value_status::Used, false,
-        "Keep SSTable index pages in the global cache after a SSTable read. Expected to improve performance for workloads with big partitions, but may degrade performance for workloads with small partitions.")
+        "Keep SSTable index pages in the global cache after a SSTable read. Expected to improve performance for workloads with big partitions, but may degrade performance for workloads with small partitions. The amount of memory usable by index cache is limited with `index_cache_fraction`.")
+    , index_cache_fraction(this, "index_cache_fraction", liveness::LiveUpdate, value_status::Used, 1.0,
```
0.2

avikivity commented on row_cache.cc:
```diff
@@ -51,17 +52,34 @@ row_cache::create_underlying_reader(read_context& ctx, mutation_source& src, con
+    size_t space_used_by_index_cache();
+}
+
+cache_tracker::cache_tracker(utils::updateable_value<double> index_cache_fraction, mutation_application_stats& app_stats, register_metrics with_metrics)
```
How about passing some struct here that contains index_cache_fraction and an std::function to determine the space used by the index cache, then we avoid the dependency (which is likely circular).

avikivity replied:
Yes, it's important to me not to reduce modularity.

michoecho replied:
Solved by moving the ownership of index cache metrics to cache_tracker.

avikivity commented on row_cache.cc:
```diff
@@ -78,7 +96,35 @@ cache_tracker::cache_tracker(mutation_application_stats& app_stats, register_met
+            size_t index_cache_space = sstables::space_used_by_index_cache();
+            bool should_evict_index = index_cache_space > total_cache_space * _index_cache_fraction.get();
+
+            return _lru.evict(should_evict_index);
```
Might want to loop here to amortize the cost of the decision.

michoecho replied:
The decision should be very cheap, amortizing it isn't worth the added complexity.

avikivity replied:
It may be cheap but the gain from an eviction can be tiny, so we have a low-cost decision against a low-gain outcome.

michoecho replied:
The thing is, I expect eviction of an item to be orders of magnitude more expensive than computing the conditional. The minimum cost of a single eviction is DRAM latency (it's an uncached linked list, after all), and usually it also entails unlinking stuff from trees, destroying cells, freeing memory, etc. Compared to that, computing the conditional should be rather insignificant — it's a few cacheable loads and a bit of arithmetic.

avikivity replied:
Yes probably true re. the uncacheability and its dominance.

avikivity commented:
Need a test.

michoecho commented:
> Need a test.

What kind of test do you want?

An integration test that spams the worst-case workload (a schema with only partition keys) and checks that the index cache stays at the defined limit?
A unit test for `lru`?
A unit test for `cache_tracker`?
All of the above?

I'm very bad at deciding what tests are worth writing. I wish we had some methodology which makes that decision for us.

tgrabiec commented on sstables/sstables.cc:
```diff
@@ -2817,6 +2817,10 @@ thread_local cached_file::metrics index_page_cache_metrics;
 static thread_local seastar::metrics::metric_groups metrics;
 
+size_t space_used_by_index_cache() {
+    return partition_index_cache::shard_stats().used_bytes + index_page_cache_metrics.cached_bytes;
```
Maybe a cleaner way to achieve that would be to have a notion of region zones. When you create caches, instead of passing `get_cache_tracker().region()` you would get `get_cache_tracker().get_index_zone()` which returns 

```
struct region_zone {
    region& r;
    allocator& alloc;
}
```

And alloc is a zone-specific wrapper around region allocator which accounts memory allocated through the zone, like the current `measuring_allocator`.


In caches, instead of:
```
with_allocator(_region.allocator(), [&] {
``` 

you'd do:
```
with_allocator(_zone.allocator(), [&] {
```

mykaul commented:
Please add an issue so we can have 'fixes' or 'refs' comment refer to it.

mykaul commented:
@michoecho - there are some conflicts that need to be resolved (seem benign).

michoecho commented:
> @michoecho - there are some conflicts that need to be resolved (seem benign).

I know. This PR wasn't just waiting on conflict resolution — it was waiting for me to fix the reliance on globals and to add tests.

michoecho commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3384/

An accident during final rebase, fixed.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3392/

tgrabiec commented on test/boost/cache_algorithm_test.cc:
```diff
@@ -0,0 +1,152 @@
+        // Sanity check. The test assumes that the total table size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(pk_number * ck_number * v_size, 2 * seastar::memory::stats().total_memory());
+
+        // A bijection between uint64_t and blobs of size pk_size.
```
There's no pk_size here.

michoecho replied:
Right, it should be `x`, not `pk_size`.

tgrabiec commented on test/boost/cache_algorithm_test.cc:
```diff
@@ -0,0 +1,152 @@
+        };
+
+        // Populate the table.
+        auto pk = bytes(sizeof(uint64_t), 0);
```
pk is unused?

michoecho replied:
Right, this line is leftover from an earlier iteration of the test.

tgrabiec reviewed (commented):
LGTM, just some nits

michoecho commented on test/boost/cache_algorithm_test.cc:
```diff
@@ -0,0 +1,152 @@
+
+        constexpr uint64_t hot_subset_size = 1000;
+        // Sanity check. The test assumes that the total *hot* index size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(hot_subset_size * pk_size * 2000, 2 * seastar::memory::stats().total_memory());
```
2000 is the `data file size : summary file size` ratio. Hence, in a table containing only primary keys, `pk_size * 2000` is approximately the size of an index page.

The sanity check here is that the maximum total size of the touched index pages is much greater than RAM. ("Maximum", because multiple rows can land in the same index page).

I'll add a comment which explains that.

michoecho commented:
v3:
- Addressed @tgrabiec's comments
- Addressed the CI failure. It turns out that reads from `system_auth.roles` were generating unpredicted cache misses, breaking the sanity checks in the test. I dealt with that by effectively disabling CQL permission refreshes (by increasing their period to a giant value).
- Doubled the memory of the test and the size of the dataset. With 128 MiB I saw some spurious bad_allocs when running the test with a large repeat count.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3415/

michoecho commented:
CI corrections:
- Added `#include <cstdint>` to the new headers, to make them self-sufficient.
- Disabled the new tests in debug mode, because debug mode ignores `--memory`. (And also the tests are relatively slow).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3418/
